src/pyodide.js

```diff
@@ -171,102 +161,127 @@ var languagePluginLoader = new Promise((resolve, reject) => {
-              resolve(resolveMsg);
-            });
-          } else {
-            console.log(resolveMsg);
-            resolve(resolveMsg);
-          }
-        }
-      };
+    // Wait for all scripts to be loaded first. Then all the run dependencies
+    // have been added. Then wait for the dependencies to be removed.
+    let successPromise = Promise.all(scriptPromises).then(() => runDepPromise);
+    try {
+      await Promise.race([ successPromise, windowErrorPromise ]);
+    } finally {
+      delete self.pyodide._module.monitorRunDependencies;
+      self.removeEventListener('error', windowErrorHandler);
+    }
 
-      // Add a handler for any exceptions that are thrown in the process of
-      // loading a package
-      var windowErrorHandler = (err) => {
-        delete self.pyodide._module.monitorRunDependencies;
-        self.removeEventListener('error', windowErrorHandler);
-        // Set up a new Promise chain, since this one failed
-        loadPackagePromise = new Promise((resolve) => resolve());
-        reject(err.message);
-      };
-      self.addEventListener('error', windowErrorHandler);
+    let packageList = [];
+    for (let pkg in toLoad) {
+      self.pyodide.loadedPackages[pkg] = toLoad[pkg];
+      packageList.push(pkg);
+    }
 
-      for (let pkg in toLoad) {
-        let scriptSrc;
-        let package_uri = toLoad[pkg];
-        if (package_uri == 'default channel') {
-          scriptSrc = `${baseURL}${pkg}.js`;
-        } else {
-          scriptSrc = `${package_uri}`;
-        }
-        _messageCallback(`Loading ${pkg} from ${scriptSrc}`)
-        loadScript(scriptSrc, () => {}, () => {
-          // If the package_uri fails to load, call monitorRunDependencies twice
-          // (so packageCounter will still hit 0 and finish loading), and remove
-          // the package from toLoad so we don't mark it as loaded, and remove
-          // the package from packageList so we don't say that it was loaded.
-          _errorCallback(`Couldn't load package from URL ${scriptSrc}`);
-          delete toLoad[pkg];
-          let packageListIndex = packageList.indexOf(pkg);
-          if (packageListIndex !== -1) {
-            packageList.splice(packageListIndex, 1);
-          }
-          for (let i = 0; i < 2; i++) {
-            self.pyodide._module.monitorRunDependencies();
-          }
-        });
-      }
+    let resolveMsg = `Loaded `;
+    if (packageList.length > 0) {
+      resolveMsg += packageList.join(', ');
+    } else {
+      resolveMsg += 'no packages'
+    }
 
-      // We have to invalidate Python's import caches, or it won't
-      // see the new files. This is done here so it happens in parallel
-      // with the fetching over the network.
-      self.pyodide.runPython('import importlib as _importlib\n' +
-                             '_importlib.invalidate_caches()\n');
-    });
+    if (!isFirefox) {
+      await preloadWasm();
+    }
+    console.log(resolveMsg);
```

Comment:
Another place we unconditionally write `info` level messages to the console.

Reply:
As I mentioned in the OP, I think the right way out is to let user supply loggers. Before that, I'm maintaining the original behaviour...